# Toggle visibility of the menubar
# Available in the View submenu or by pressing shift+F12 in the guest.
# http://paranoid-terminus.org/confined-application-with-qemu.html#borderless
--- a/tools/spicy.c
+++ b/tools/spicy.c
@@ -460,6 +460,15 @@
     g_key_file_set_boolean(keyfile, "ui", "toolbar", state);
 }
 
+static void menu_cb_menubar(GtkToggleAction *action, gpointer data)
+{
+    SpiceWindow *win = data;
+    gboolean state = gtk_toggle_action_get_active(action);
+
+    gtk_widget_set_visible(win->menubar, state);
+    g_key_file_set_boolean(keyfile, "ui", "menubar", state);
+}
+
 static void menu_cb_statusbar(GtkToggleAction *action, gpointer data)
 {
     SpiceWindow *win = data;
@@ -522,6 +531,9 @@
             toggle = gtk_action_group_get_action(win->ag, "Toolbar");
             state = gtk_toggle_action_get_active(GTK_TOGGLE_ACTION(toggle));
             gtk_widget_set_visible(win->toolbar, state);
+	    toggle = gtk_action_group_get_action(win->ag, "Menubar");
+            state = gtk_toggle_action_get_active(GTK_TOGGLE_ACTION(toggle));
+            gtk_widget_set_visible(win->menubar, state);
             toggle = gtk_action_group_get_action(win->ag, "Statusbar");
             state = gtk_toggle_action_get_active(GTK_TOGGLE_ACTION(toggle));
             gtk_widget_set_visible(win->statusbar, state);
@@ -683,6 +695,11 @@
         gtk_widget_set_visible(win->toolbar, state);
     g_clear_error(&error);
 
+    state = g_key_file_get_boolean(keyfile, "ui", "menubar", &error);
+    if (error == NULL)
+        gtk_widget_set_visible(win->menubar, state);
+    g_clear_error(&error);
+
     state = g_key_file_get_boolean(keyfile, "ui", "statusbar", &error);
     if (error == NULL)
         gtk_widget_set_visible(win->statusbar, state);
@@ -847,6 +864,11 @@
         .label       = "Statusbar",
         .callback    = G_CALLBACK(menu_cb_statusbar),
     },{
+        .name        = "Menubar",
+        .label       = "Menubar",
+        .callback    = G_CALLBACK(menu_cb_menubar),
+        .accelerator = "<shift>F12",
+    },{
         .name        = "Toolbar",
         .label       = "Toolbar",
         .callback    = G_CALLBACK(menu_cb_toolbar),
@@ -923,6 +945,7 @@
 "    <menu action='ViewMenu'>\n"
 "      <menuitem action='Fullscreen'/>\n"
 "      <menuitem action='Toolbar'/>\n"
+"      <menuitem action='Menubar'/>\n"
 "      <menuitem action='Statusbar'/>\n"
 "    </menu>\n"
 "    <menu action='InputMenu'>\n"
@@ -1196,6 +1219,10 @@
     state = gtk_widget_get_visible(win->toolbar);
     gtk_toggle_action_set_active(GTK_TOGGLE_ACTION(toggle), state);
 
+    toggle = gtk_action_group_get_action(win->ag, "Menubar");
+    state = gtk_widget_get_visible(win->menubar);
+    gtk_toggle_action_set_active(GTK_TOGGLE_ACTION(toggle), state);
+
     toggle = gtk_action_group_get_action(win->ag, "Statusbar");
     state = gtk_widget_get_visible(win->statusbar);
     gtk_toggle_action_set_active(GTK_TOGGLE_ACTION(toggle), state);

